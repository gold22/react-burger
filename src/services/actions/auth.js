export const REGISTER_USER_REQUEST = 'REGISTER_USER_REQUEST';
export const REGISTER_USER_SUCCESS = 'REGISTER_USER_SUCCESS';
export const REGISTER_USER_ERROR = 'REGISTER_USER_ERROR';

export const registerUser = (user) => async (dispatch, getState, apiClient) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result.user,
            accessToken: result.accessToken,
            refreshToken: result.refreshToken,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: error.message,
        });
    }
};
