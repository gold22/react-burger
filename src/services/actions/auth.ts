import {
    REGISTER_USER_REQUEST,
    REGISTER_USER_SUCCESS,
    REGISTER_USER_ERROR,
    LOGIN_USER_REQUEST,
    LOGIN_USER_SUCCESS,
    LOGIN_USER_ERROR,
    LOGOUT_USER_REQUEST,
    LOGOUT_USER_SUCCESS,
    LOGOUT_USER_ERROR,
    GET_USER_REQUEST,
    GET_USER_SUCCESS,
    GET_USER_ERROR,
    UPDATE_USER_REQUEST,
    UPDATE_USER_SUCCESS,
    UPDATE_USER_ERROR,
} from '../constants/auth';
import { TApiUser, TApiUserCredentials } from '../types/api';
import { TDispatch, TGetState, TThunk } from '../types/store';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';

export const registerUser: TThunk = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: REGISTER_USER_REQUEST,
    });
    try {
        const result = await apiClient.registerUser(user);
        dispatch({
            type: REGISTER_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: REGISTER_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logInUser: TThunk = (user: TApiUserCredentials) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGIN_USER_REQUEST,
    });
    try {
        const result = await apiClient.logInUser(user);
        dispatch({
            type: LOGIN_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: LOGIN_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const logOutUser: TThunk = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: LOGOUT_USER_REQUEST,
    });
    try {
        await apiClient.logOutUser();
        dispatch({
            type: LOGOUT_USER_SUCCESS,
        });
    } catch (error) {
        dispatch({
            type: LOGOUT_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const getUser = () => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_USER_REQUEST,
    });
    try {
        const user = await apiClient.getUser();
        dispatch({
            type: GET_USER_SUCCESS,
            user,
        });
    } catch (error) {
        dispatch({
            type: GET_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};

export const updateUser = (user: TApiUser) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: UPDATE_USER_REQUEST,
    });
    try {
        const result = await apiClient.updateUser(user);
        dispatch({
            type: UPDATE_USER_SUCCESS,
            user: result,
        });
    } catch (error) {
        dispatch({
            type: UPDATE_USER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
