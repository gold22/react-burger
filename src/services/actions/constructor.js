import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';

export const setBun = (bun) => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient, index) => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index) => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = () => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex, newIndex) => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
