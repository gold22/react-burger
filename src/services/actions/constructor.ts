import { v4 as uuid } from 'uuid';
import {
    SET_BUN,
    ADD_INGREDIENT,
    REMOVE_INGREDIENT,
    REMOVE_INGREDIENTS,
    MOVE_INGREDIENT,
} from '../constants/constructor';
import { TIngredient } from '../types';
import {
    IAddIngredientAction,
    IMoveIngredientAction,
    IRemoveIngredientAction,
    IRemoveIngredientsAction,
    ISetBunAction,
} from '../types/actions/constructor';

export const setBun = (bun: TIngredient): ISetBunAction => ({
    type: SET_BUN,
    bun,
});

export const addIngredient = (ingredient: TIngredient, index: number): IAddIngredientAction => ({
    type: ADD_INGREDIENT,
    ingredient: {
        ...ingredient,
        uuid: uuid(),
    },
    index,
});

export const removeIngredient = (index: number): IRemoveIngredientAction => ({
    type: REMOVE_INGREDIENT,
    index,
});

export const removeIngredients = (): IRemoveIngredientsAction => ({
    type: REMOVE_INGREDIENTS,
});

export const moveIngredient = (oldIndex: number, newIndex: number): IMoveIngredientAction => ({
    type: MOVE_INGREDIENT,
    oldIndex,
    newIndex,
});
