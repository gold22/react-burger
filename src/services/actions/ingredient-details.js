import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
} from '../constants/ingredient-details';

export const showIngredientDetails = (ingredient) => ({
    type: SHOW_INGREDIENT_DETAILS,
    ingredient,
});

export const hideIngredientDetails = () => ({
    type: HIDE_INGREDIENT_DETAILS,
});
