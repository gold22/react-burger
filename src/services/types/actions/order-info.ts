import {
    GET_ORDER_REQUEST,
    GET_ORDER_SUCCESS,
    GET_ORDER_ERROR,
} from '../../constants/order-info';
import { TOrder } from '../index';

export interface IGetOrderAction {
    readonly type: typeof GET_ORDER_REQUEST;
}

export interface IGetOrderSuccessAction {
    readonly type: typeof GET_ORDER_SUCCESS;
    readonly order: TOrder;
}

export interface IGetOrderFailedAction {
    readonly type: typeof GET_ORDER_ERROR;
    readonly message: string;
}

export type TGetOrderActions =
    | IGetOrderAction
    | IGetOrderSuccessAction
    | IGetOrderFailedAction;
