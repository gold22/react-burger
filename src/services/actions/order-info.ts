import {
    GET_ORDER_REQUEST,
    GET_ORDER_SUCCESS,
    GET_ORDER_ERROR,
    CLEAR_ORDER,
} from '../constants/order-info';
import { TDispatch, TGetState, TThunk } from '../types/store';
import { IClearOrderAction } from '../types/actions/order-info';
import ApiClient from '../api-client';
import { getErrorMessage } from '../../utils/error';
import { mapOrder } from '../../utils/orders';

export const clearOrder = (): IClearOrderAction => ({
    type: CLEAR_ORDER,
});

export const getOrder: TThunk = (orderNumber: string) => async (
    dispatch: TDispatch,
    getState: TGetState,
    apiClient: ApiClient,
) => {
    dispatch({
        type: GET_ORDER_REQUEST,
    });
    try {
        const order = await apiClient.getOrder(orderNumber);
        dispatch({
            type: GET_ORDER_SUCCESS,
            order: mapOrder(order),
        });
    } catch (error) {
        dispatch({
            type: GET_ORDER_ERROR,
            message: getErrorMessage(error),
        });
    }
};
