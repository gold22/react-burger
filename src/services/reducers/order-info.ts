import {
    GET_ORDER_REQUEST,
    GET_ORDER_SUCCESS,
    GET_ORDER_ERROR,
} from '../constants/order-info';
import { TOrder } from '../types';
import { TGetOrderActions } from '../types/actions/order-info';

export type TOrderInfoState = {
    order: TOrder | null;
    isLoading: boolean;
    loadError: string | null;
};

const initialState: TOrderInfoState = {
    order: null,
    isLoading: true,
    loadError: null,
};

export const orderInfoReducer = (
    // eslint-disable-next-line @typescript-eslint/default-param-last
    state: TOrderInfoState = initialState,
    action: TGetOrderActions,
): TOrderInfoState => {
    switch (action.type) {
    case GET_ORDER_REQUEST: {
        return {
            ...state,
            isLoading: true,
        };
    }
    case GET_ORDER_SUCCESS: {
        return {
            ...state,
            isLoading: false,
            loadError: null,
            order: action.order,
        };
    }
    case GET_ORDER_ERROR: {
        return {
            ...initialState,
            isLoading: false,
            loadError: action.message,
        };
    }
    default: {
        return state;
    }
    }
};
