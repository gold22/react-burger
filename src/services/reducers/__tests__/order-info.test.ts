import { orderInfoReducer, TOrderInfoState } from '../order-info';
import {
    CLEAR_ORDER,
    GET_ORDER_ERROR,
    GET_ORDER_REQUEST,
    GET_ORDER_SUCCESS,
} from '../../constants/order-info';
import { order9932 } from '../../../__fixtures__/orders';
import { IGetOrderSuccessAction } from '../../types/actions/order-info';

const initialState: TOrderInfoState = orderInfoReducer(undefined, {} as any);

describe('order info reducer', () => {
    it('should return the initial state', () => {
        expect(initialState).toMatchSnapshot();
    });

    it('should handle GET_ORDER_REQUEST', () => {
        expect(orderInfoReducer(undefined, { type: GET_ORDER_REQUEST }))
            .toMatchSnapshot();
    });

    it('should handle GET_ORDER_SUCCESS', () => {
        const action: IGetOrderSuccessAction = {
            type: GET_ORDER_SUCCESS,
            order: order9932,
        };
        const state: TOrderInfoState = {
            ...initialState,
            isLoading: true,
        };
        expect(orderInfoReducer(state, action))
            .toMatchSnapshot();
    });

    it('should handle GET_ORDER_ERROR', () => {
        const state: TOrderInfoState = {
            ...initialState,
            isLoading: true,
        };
        expect(orderInfoReducer(state, { type: GET_ORDER_ERROR, message: 'Get order error' }))
            .toMatchSnapshot();
    });

    it('should handle CLEAR_ORDER', () => {
        const state: TOrderInfoState = {
            ...initialState,
            order: order9932,
        };
        expect(orderInfoReducer(state, { type: CLEAR_ORDER }))
            .toMatchSnapshot();
    });
});
