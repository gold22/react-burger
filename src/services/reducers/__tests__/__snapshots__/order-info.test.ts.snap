// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order info reducer should handle CLEAR_ORDER 1`] = `
Object {
  "isLoading": true,
  "loadError": null,
  "order": null,
}
`;

exports[`order info reducer should handle GET_ORDER_ERROR 1`] = `
Object {
  "isLoading": false,
  "loadError": "Get order error",
  "order": null,
}
`;

exports[`order info reducer should handle GET_ORDER_REQUEST 1`] = `
Object {
  "isLoading": true,
  "loadError": null,
  "order": null,
}
`;

exports[`order info reducer should handle GET_ORDER_SUCCESS 1`] = `
Object {
  "isLoading": false,
  "loadError": null,
  "order": Object {
    "createdAt": 2022-02-12T15:47:31.913Z,
    "id": "6207d6936d7cd8001b2d5a91",
    "ingredients": Array [
      "60d3b41abdacab0026a733c6",
      "60d3b41abdacab0026a733ce",
      "60d3b41abdacab0026a733c6",
    ],
    "name": "Традиционный-галактический краторный бургер",
    "number": 9932,
    "status": "done",
    "updatedAt": 2022-02-12T15:47:32.195Z,
  },
}
`;

exports[`order info reducer should return the initial state 1`] = `
Object {
  "isLoading": true,
  "loadError": null,
  "order": null,
}
`;
